File: DesktopTools/dotNET/DesktopToolApp/DesktopToolApp/Functions/DeviceMaintenance/DeviceTimestamp.cs
using AppCommon;
using System;
using System.Text;
using static DesktopTool.BLEDefines;
using static DesktopTool.FunctionDefines;
using static DesktopTool.FunctionMessage;

namespace DesktopTool
{
    internal class DeviceTimestamp
    {
        // プロパティー
        public string CurrentTimestampString { get; private set; }
        public string CurrentTimestampLogString { get; private set; }
        private string FunctionName { get; set; }

        public DeviceTimestamp()
        {
            CurrentTimestampString = string.Empty;
            CurrentTimestampLogString = string.Empty;
            FunctionName = string.Empty;
        }

        //
        // 現在時刻照会処理
        //
        public delegate void NotifyResponseQueryHandler(DeviceTimestamp sender, bool success, string errorMessage);
        private event NotifyResponseQueryHandler NotifyResponseQuery = null!;

        public void Inquiry(NotifyResponseQueryHandler notifyResponseQueryHandler)
        {
            // コールバックを設定
            NotifyResponseQuery += notifyResponseQueryHandler;

            // 機能名を設定
            FunctionName = nameof(Inquiry);

            // BLEデバイスに接続
            new BLEU2FTransport().Connect(OnNotifyConnection, U2F_BLE_SERVICE_UUID_STR);
        }

        private void OnNotifyConnection(BLETransport sender, bool success, string errorMessage)
        {
            if (success == false) {
                // 失敗時は上位クラスに制御を戻す
                NotifyResponseQuery?.Invoke(this, false, errorMessage);
                NotifyResponseQuery = null!;
                return;
            }

            // コールバックを登録
            sender.RegisterResponseReceivedHandler(ResponseReceivedHandler);

            if (FunctionName.Equals(nameof(Update))) {
                // 現在時刻設定コマンドを実行
                PerformUpdateCommand(sender);

            } else {
                // 現在時刻参照コマンドを実行
                PerformInquiryCommand(sender);
            }
        }

        private void OnResponseInquiryCommand(BLETransport sender, byte[] responseBytes)
        {
            // 管理ツールの現在時刻を取得
            string toolTimestamp = DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss");

            // 現在時刻文字列はレスポンスの２バイト目から19文字
            byte[] data = AppUtil.ExtractCBORBytesFromResponse(responseBytes);

            // デバイスの現在時刻文字列
            string deviceTimestamp = Encoding.UTF8.GetString(data);

            // 現在時刻文字列をログ出力
            CurrentTimestampLogString = string.Format(MSG_DEVICE_TIMESTAMP_CURRENT_DATETIME_LOG_FORMAT, toolTimestamp, deviceTimestamp);

            // 現在時刻文字列を画面表示
            CurrentTimestampString = string.Format(MSG_DEVICE_TIMESTAMP_CURRENT_DATETIME_FORMAT, toolTimestamp, deviceTimestamp);

            // 上位クラスに制御を戻す
            TerminateCommand(sender, true, string.Empty);
        }

        //
        // 現在時刻設定処理
        //
        public void Update(NotifyResponseQueryHandler notifyResponseQueryHandler)
        {
            // コールバックを設定
            NotifyResponseQuery += notifyResponseQueryHandler;

            // 機能名を設定
            FunctionName = nameof(Update);

            // BLEデバイスに接続
            new BLEU2FTransport().Connect(OnNotifyConnection, U2F_BLE_SERVICE_UUID_STR);
        }

        //
        // コールバック関数
        //
        private void ResponseReceivedHandler(BLETransport sender, bool success, string errorMessage, byte responseCMD, byte[] responseBytes)
        {
            // レスポンス受信失敗時はエラー扱い
            if (success == false) {
                TerminateCommand(sender, false, errorMessage);
                return;
            }

            // レスポンスステータスをチェック
            byte status = responseBytes[0];
            if (status != CTAP1_ERR_SUCCESS) {
                TerminateCommand(sender, false, string.Format(MSG_FORMAT_OCCUR_UNKNOWN_ERROR_ST, status));
                return;
            }

            OnResponseInquiryCommand(sender, responseBytes);
        }

        //
        // 内部処理
        //
        private void PerformInquiryCommand(BLETransport sender)
        {
            // 現在時刻参照コマンドを実行
            sender.SendRequest(U2F_COMMAND_MSG, new byte[] { VENDOR_COMMAND_GET_TIMESTAMP });
        }

        private void PerformUpdateCommand(BLETransport sender)
        {
            // 現在のUNIX時刻を取得
            TimeSpan t = DateTime.UtcNow - new DateTime(1970, 1, 1);
            UInt32 nowEpochSeconds = (UInt32)t.TotalSeconds;

            // 現在時刻設定用のリクエストデータを生成
            byte[] data = new byte[] { VENDOR_COMMAND_SET_TIMESTAMP, 0x00, 0x00, 0x00, 0x00 };
            AppUtil.ConvertUint32ToBEBytes(nowEpochSeconds, data, 1);

            // 現在時刻参照コマンドを実行
            sender.SendRequest(U2F_COMMAND_MSG, data);
        }

        //
        // 終了処理
        //
        private void TerminateCommand(BLETransport sender, bool success, string message)
        {
            // コールバックを解除
            sender.UnregisterResponseReceivedHandler(ResponseReceivedHandler);

            // 切断処理
            sender.Disconnect();

            // 上位クラスに制御を戻す
            NotifyResponseQuery?.Invoke(this, success, message);
            NotifyResponseQuery = null!;
        }
    }
}

File: DesktopTools/macOS/Functions/DeviceMaintenance/DeviceTimestamp.m
//
//  DeviceTimestamp.m
//  DesktopTool
//
//  Created by Makoto Morita on 2024/01/26.
//
#import "AppUtil.h"
#import "BLEU2FTransport.h"
#import "DeviceTimestamp.h"
#import "FunctionDefine.h"
#import "FunctionMessage.h"

@interface DeviceTimestamp () <BLETransportDelegate>
    // 上位クラスの参照を保持
    @property (nonatomic) id                            delegate;
    @property (nonatomic) BLEU2FTransport              *transport;
    // 実行機能を保持
    @property (nonatomic) NSString                     *functionName;
    // PCの現在時刻を保持
    @property (nonatomic) NSString                     *toolTimestamp;
    // デバイスの現在時刻文字列を保持
    @property (nonatomic) NSString                     *deviceTimestamp;

@end

@implementation DeviceTimestamp

    - (instancetype)initWithDelegate:(id)delegate {
        self = [super init];
        if (self) {
            [self setDelegate:delegate];
            [self setTransport:[[BLEU2FTransport alloc] initWithDelegate:self]];
        }
        return self;
    }

    - (void)BLETransport:(BLETransport *)bleTransport didUpdateState:(bool)available {
        [[self delegate] DeviceTimestamp:self didUpdateState:available];
    }

    - (void)inquiry {
        // U2F BLEサービスに接続
        [self setFunctionName:NSStringFromSelector(_cmd)];
        [[self transport] transportWillConnect];
    }

    - (void)update {
        // U2F BLEサービスに接続
        [self setFunctionName:NSStringFromSelector(_cmd)];
        [[self transport] transportWillConnect];
    }

    - (void)BLETransport:(BLETransport *)bleTransport didConnect:(bool)success withErrorMessage:(NSString *)errorMessage {
        if (success == false) {
            // U2F BLEサービスに接続失敗時
            [[self delegate] DeviceTimestamp:self didNotifyResponseQuery:false withErrorMessage:errorMessage];
            return;
        }
        if ([[self functionName] isEqualToString:@"update"]) {
            // 現在時刻設定コマンドを実行
            [self performUpdateCommand:bleTransport];
        } else {
            // 現在時刻参照コマンドを実行
            [self performInquiryCommand:bleTransport];
        }
    }

    - (void)BLETransport:(BLETransport *)bleTransport didReceiveResponse:(bool)success withErrorMessage:(NSString *)errorMessage withCMD:(uint8_t)responseCMD withData:(NSData *)responseData {
        if (success == false) {
            // コマンド受信失敗時はログ出力
            [self disconnectAndTerminateCommand:bleTransport withSuccess:false withErrorMessage:errorMessage];
            return;
        }
        // レスポンスデータを抽出
        uint8_t *responseBytes = (uint8_t *)[responseData bytes];
        // ステータスをチェック
        uint8_t status = responseBytes[0];
        if (status != CTAP1_ERR_SUCCESS) {
            NSString *statusErrorMessage = [NSString stringWithFormat:MSG_FORMAT_OCCUR_UNKNOWN_ERROR_ST, status];
            [self disconnectAndTerminateCommand:bleTransport withSuccess:false withErrorMessage:statusErrorMessage];
            return;
        }
        // レスポンスデータから現在時刻を抽出
        [self BLETransport:bleTransport didResponseInquiryCommand:responseData];
    }

    - (void)disconnectAndTerminateCommand:(BLETransport *)bleTransport withSuccess:(bool)success withErrorMessage:(NSString *)errorMessage {
        // BLE接続を切断し、制御を戻す
        [bleTransport transportWillDisconnect];
        [[self delegate] DeviceTimestamp:self didNotifyResponseQuery:success withErrorMessage:errorMessage];
    }

#pragma mark - 現在時刻参照

    - (void)performInquiryCommand:(BLETransport *)bleTransport {
        // 現在時刻参照コマンドを実行
        uint8_t requestBytes[] = {VENDOR_COMMAND_GET_TIMESTAMP};
        NSData *requestData = [[NSData alloc] initWithBytes:requestBytes length:sizeof(requestBytes)];
        [bleTransport transportWillSendRequest:U2F_COMMAND_MSG withData:requestData];
    }

    - (void)BLETransport:(BLETransport *)bleTransport didResponseInquiryCommand:(NSData *)responseData {
        // PCの現在時刻を取得
        NSDateFormatter *df = [[NSDateFormatter alloc] init];
        [df setLocale:[[NSLocale alloc] initWithLocaleIdentifier:@"ja_JP"]];
        [df setDateFormat:@"yyyy/MM/dd HH:mm:ss"];
        [self setToolTimestamp:[df stringFromDate:[NSDate date]]];
        // 現在時刻文字列はレスポンスの２バイト目から19文字
        char timestampString[20];
        size_t lastPos = sizeof(timestampString) - 1;
        memcpy(timestampString, [responseData bytes] + 1, lastPos);
        timestampString[lastPos] = 0;
        // デバイスの現在時刻文字列
        [self setDeviceTimestamp:[[NSString alloc] initWithUTF8String:timestampString]];
        // 上位クラスに制御を戻す
        [self disconnectAndTerminateCommand:bleTransport withSuccess:true withErrorMessage:nil];
    }

    - (NSString *)currentTimestampLogString {
        // ログ出力用の現在時刻文字列を生成
        NSString *str = [NSString stringWithFormat:MSG_DEVICE_TIMESTAMP_CURRENT_DATETIME_LOG_FORMAT, [self toolTimestamp], [self deviceTimestamp]];
        return str;
    }

    - (NSString *)currentTimestampString {
        // 画面表示用の現在時刻文字列を生成
        NSString *str = [NSString stringWithFormat:MSG_DEVICE_TIMESTAMP_CURRENT_DATETIME_FORMAT, [self toolTimestamp], [self deviceTimestamp]];
        return str;
    }

#pragma mark - 現在時刻設定

    - (void)performUpdateCommand:(BLETransport *)bleTransport {
        // 現在のUNIX時刻を取得
        NSDate *now = [NSDate date];
        NSTimeInterval nowEpochSeconds = [now timeIntervalSince1970];
        // 現在時刻設定用のリクエストデータを生成
        uint8_t requestBytes[] = {VENDOR_COMMAND_SET_TIMESTAMP, 0x00, 0x00, 0x00, 0x00};
        [AppUtil convertUint32:(uint32_t)nowEpochSeconds toBEBytes:(requestBytes + 1)];
        NSData *requestData = [[NSData alloc] initWithBytes:requestBytes length:sizeof(requestBytes)];
        // 現在時刻設定コマンドを実行
        [bleTransport transportWillSendRequest:U2F_COMMAND_MSG withData:requestData];
    }

@end
